### ISSUE TITLE
The optimizer AdamW performs poor with CPU  compared to GPU in libtorch

### ISSUE THREAD

**Issue by w1005444804** [pytorch/pytorch#117940]
### 🐛 Describe the bug

When I use CPU device to fine tune a segment-model and set the optimizer:  adamw, there will be a significant accuracy error in the early stages of training; But when using GPU device, it is not easy or difficult to detect such a large error. Using SGD doesn't have such a problem。

### Versions

libtorch: 2.1.2+cu121 and 2.1.2+cpu  

cc @jbschlosser @vincentqb @albanD @janeyx99 @crcrpar @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

**Comment by janeyx99** [comment]
hello! If I understand this issue correctly, are you comparing AdamW on CPU vs GPU? If so, how big are you finding the difference? It is normal for the CPU and CUDA implementations to diverge a little bit corresponding to dtype, see https://pytorch.org/docs/stable/notes/numerical_accuracy.html. With AdamW, since it boils down to multiple math operations in sequence, the divergence can increase in just one step. For SGD, there are much fewer ops (just one if you don’t use momentum), so the difference is less telling.

**Comment by w1005444804** [comment]
Hi, @janeyx99, yes,Let me show you two pictures.
![微信图片_20240123094148](https://github.com/pytorch/pytorch/assets/6340026/41a4226a-ec73-43b8-a63e-9f28175d917a)
![微信图片_20240123094157](https://github.com/pytorch/pytorch/assets/6340026/1ebb0ed3-b78b-4290-9173-4ce59bebe8de)

**Comment by janeyx99** [comment]
Is `accuracy_fg` the number to look at? What is it measuring?

**Comment by w1005444804** [comment]
yes, accuracy_ fg means foreground accuracy.

**Comment by janeyx99** [comment]
I am still not clear what that measures. What mathematical value does it compare? What is the standard it is comparing to?

**Comment by w1005444804** [comment]
In fact, you don't have to know what it's measuring, it represents an accuracy rate, for comparing the values run on the GPU side, you will find that the values obtained by the CPU side training are much lower, for example, the first result is 100%, the second time is 17%, but the second result is 96% on the GPU side, sorry I don't have a good level of English.I hope that the training results on the CPU side should be similar to those on the GPU side

**Comment by janeyx99** [comment]
Without more details about how accuracy is computed, this behavior I believe is expected due to differences in weights initialization on CPU vs GPU and the numerical differences as I've mentioned earlier. If you'd like a more finegrained look into your data and weights setup, the forum may be a better place: https://discuss.pytorch.org/!

**Comment by w1005444804** [comment]
@janeyx99, You are right, thanks! there are two sets of initial parameters, one from the CPU and the other from the GPU. Although the accuracy was high during the first test, the initial parameters obtained from training the model on the CPU side are more suitable for CPU side fine-tuning model.

**Comment by mingfeima** [comment]
The implementation of adam is not exactly the same on CPU and GPU. on the GPU side it has a fused kernel `_fused_adam_`: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L15381

which is missing on CPU which means that it is a un-fused implementation. So the numerical error would vary.

**Comment by janeyx99** [comment]
closing issue as resolved!
